sdk/core/azure-core/azure/core/async_paging.py

```diff
@@ -85,26 +88,45 @@ def __init__(
          list of ReturnType
         :param str continuation_token: The continuation token needed by get_next
         """
-        self._get_next = get_next
-        self._extract_data = extract_data
+        self._initial_request = kwargs.pop("initial_request", None)
+        self._initial_response = kwargs.pop("initial_response", None)
+        if self._initial_response:
+            self._initial_request = self._initial_response.http_response.request
+
+        if get_next or extract_data:
+            if paging_method:
+                raise ValueError(
+                    "You can't pass in both a paging method and a callback for get_next or extract_data. "
+                    "We recomment you only pass in a paging method, since passing in callbacks is legacy."
+                )
+            self._paging_method = _LegacyPagingMethod(get_next, extract_data)
+        else:
+            if not self._initial_request and not self._initial_response:
+                raise ValueError(
+                    "You must either supply the initial request the paging method must call, or provide "
+                    "the initial response"
+                )
+            self._paging_method = paging_method
+            self._paging_method.initialize(**kwargs)
         self.continuation_token = continuation_token
-        self._did_a_call_already = False
-        self._response = None
-        self._current_page = None
+        self._response = None  # type: Optional[ResponseType]
+        self._current_page = None  # type: Optional[Iterable[ReturnType]]
 
     async def __anext__(self):
-        if self.continuation_token is None and self._did_a_call_already:
+        if self._paging_method.finished(self.continuation_token):
             raise StopAsyncIteration("End of paging")
         try:
-            self._response = await self._get_next(self.continuation_token)
+            try:
```

Comment:
Why nested try?

Reply:
good point, un-nested

Reply:
had to re-nest @xiangyan99 because flattening the tries doesn't allow me to enter multiple except statements. In the case of a legacy call to paging with an `AzureError`, I need to enter both the `TypeError` except, and the `AzureError` except, which won't happen unless it's nested